### PR TITLE
Make CLI package location and entrypoint location modifiable; remove ProjectStructureDetector

### LINKED ISSUE #55: Known project structures are not detected properly
When `sidekick init` is run in the `packages` directory, the project structures `ProjectStructure.rootWithPackages` and `ProjectStructure.multiPackage` aren't recognized.

## How to reproduce
1. Set up project
```bash
❯ cp -r sidekick/sidekick/test/templates/multi_package sample_project
# or cp -r sidekick/sidekick/test/templates/root_with_packages sample_project
```
2. Run `sidekick init` in the `packages` directory
```bash
❯ cd sample_project/packages
❯ sidekick init
Welcome to sidekick. You're about to initialize a sidekick project

Please select a name for your sidekick CLI.
...

Generating p_sidekick
Analyzing project structure in /Users/pepe/dev/repos/sample_project/packages
The project structure is not yet supported. Please open an issue at https://github.com/phntmxyz/sidekick/issues with details about your project structure
```

If `sidekick init` is run from `sample_project` instead of `sample_project/packages`, the project structure is detected properly and the command succeeds.

### PR DESCRIPTION
**Pull request by Giuspepe**
Fixes #55 

## Changes
- The directory in which the CLI package is created in is modifiable with the `--cliPackageDirectory` parameter (previously always was `<initDirectory>/packages`)
- The directory in which the entrypoint script is created in is modifiable with the `--entrypointDirectory` parameter (previously always was `<initDirectory>`)
- The `initDirectory` parameter can now also be passed explicity with `--initDirectory` (previous behavior of being able to pass it as an unnamed positional parameter and falling back to the current directory is preserved)
- Deleted `ProjectStructureDetector` to simplify CLI package creation. This fixes #55
- Packages for creating package getters in the CLI are found with `Repository.findAllPackages` instead of only looking at the `<initDir>/packages` directory

### CODE REVIEW

File: sidekick/lib/src/init/init_command.dart [hunk 1]
```diff
@@ -15,14 +16,37 @@ class InitCommand extends Command {
...
+    argParser.addOption(
+      'cliPackageDirectory',
+      abbr: 'c',
+      help: 'The directory in which the CLI package should be created. \n'
+          'This directory must be within the entrypointDirectory, or '
+          'if the entrypointDirectory is inside a git repository, it must be '
+          'within the git repository. \n'
+          'If this path is given as a relative path and there is a git repository, '
+          'it is resolved relative to the repository root. Else it is resolved '
+          'relative to the entrypointDirectory.',
```
Comment (passsy) [inline sidekick/lib/src/init/init_command.dart:37]: I would expect the directory to be resolved from `cwd`, not the `entrypointDirectory`

File: sidekick/lib/src/init/init_command.dart [hunk 2]
```diff
@@ -32,29 +56,61 @@ class InitCommand extends Command {
...
+
+    final mainProjectPath = argResults!['mainProjectPath'] as String?;
+    DartPackage? mainProject = mainProjectPath != null
+        ? DartPackage.fromDirectory(
+            repoRoot.resolveAbsoluteOrRelativeDirPath(mainProjectPath),
+          )
+        :
+        // TODO(pepe): not sure if repoRoot ?? entrypointDir or entrypointDir ?? repoRoot is better - or just null if mainProjectPath is unspecified
+        (DartPackage.fromDirectory(entrypointDir) ??
+            DartPackage.fromDirectory(repoRoot));
```
Comment (passsy) [inline sidekick/lib/src/init/init_command.dart:99]: `entrypointDir ?? repoRoot` 👍

File: sidekick/lib/src/util/directory_extension.dart [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    return this.isWithin(dir) ||
+        // canonicalize is necessary, otherwise '/a/b/c' != '/a/b/c/' != '/a/b/c/.' != '/a/b/c/../c'
+        dir.canonicalized.path == canonicalized.path;
+  }
+
+  /// When [path] is absolute, returns the directory at that path.
+  /// Else resolves the [path] relative to this directory.
+  Directory resolveAbsoluteOrRelativeDirPath(String path) =>
+      (Directory(path).isAbsolute ? Directory(path) : directory(path))
+          .canonicalized;
```
Comment (passsy) [inline sidekick/lib/src/util/directory_extension.dart:17]: `canonicalize` always converts paths to absolute paths. This means this method works not different from directly calling `canonicalize`

https://github.com/dart-lang/path/blob/9955b27b9bb98d87591208e19eb01c51d29fd467/lib/src/context.dart#L344-L351

Comment (Giuspepe) [reply sidekick/lib/src/util/directory_extension.dart:17]: If a relative path is given to `canonicalize`, it resolves it relative to the current working directory. This method resolves it relative to the directory it is called on - not necessarily on the current working directory.

```dart
    final resolveFromDir = Directory('/a/b/c');

    // path is relative
    print(resolveFromDir.resolveAbsoluteOrRelativeDirPath('d/e/f').path);
          // -> /a/b/c/d/e/f
    print(canonicalize('d/e/f'));
          // -> <Directory.current.path>/d/e/f

    // path is absolute
    print(resolveFromDir.resolveAbsoluteOrRelativeDirPath('/d/e/f').path);
          // -> /d/e/f
    print(canonicalize('/d/e/f'));
          // d/e/f
```

But I found `Context(current: resolveRelativeToThisDir).canonicalize` which achieves the same as this method!

File: sidekick/lib/src/init/init_command.dart (hunk 1 above)
Comment (Giuspepe) [reply sidekick/lib/src/init/init_command.dart:37]: Which one of these options do you mean?

Resolve relative paths ...
1. only relative to `cwd`
2. relative to `repoRoot` and if that doesn't work relative to `cwd`

Comment (passsy) [reply sidekick/lib/src/init/init_command.dart:38]: always relative to cwd

**Pull request merged**